server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -310,6 +311,49 @@ public void parse(ParseContext context) throws IOException {
      */
     protected abstract void parseCreateField(ParseContext context) throws IOException;
 
+    /**
+     * Given access to a document's _source, return this field's values.
+     *
+     * In addition to pulling out the values, mappers can parse them into a standard form. This
+     * method delegates parsing to {@link #parseSourceValue} for parsing. Most mappers will choose
+     * to override {@link #parseSourceValue} -- for example numeric field mappers make sure to
+     * parse the  source value into a number of the right type.
+     *
+     * Some mappers may need more flexibility and can override this entire method instead.
```

Comment:
I've been *trying* to declare methods as either `final` or `abstract` mostly because I keep bumping into situations where things that override "big" methods do a lot of copy and pasting or are just otherwise confusing. You certainly don't have to do it, but if you can build this so it is `final` and things that need to customize the logic more can do so via hooks or something I'd feel better.

Reply:
I'm in agreement, this structure felt tricky to me. For context, the only field mapper that will override `lookupValues` directly is `ConstantKeywordFieldMapper`.

I played around with several alternatives but couldn't find a better approach. Do you have any suggestions (I'm wondering how we could be able to make this `final`)?

Reply:
I'm not always a huge fan of extension through inheriance, but maybe this should be implemented in a `FieldMapper` that `ConstantKeywordFieldMapper` doesn't extend from and everything else does? I dunno.

Reply:
I have no idea if it is better, but it is a thing I might try